dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt

```diff
@@ -0,0 +1,144 @@
+package org.jetbrains.kotlinx.dataframe.io
+
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.arrow.vector.ipc.ArrowFileWriter
+import org.apache.arrow.vector.ipc.ArrowStreamWriter
+import org.apache.arrow.vector.types.pojo.Schema
+import org.jetbrains.kotlinx.dataframe.AnyFrame
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.slf4j.LoggerFactory
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.io.OutputStream
+import java.nio.channels.Channels
+import java.nio.channels.WritableByteChannel
+
+public val ignoreMismatchMessage: (ConvertingMismatch) -> Unit = { message: ConvertingMismatch -> }
+public val writeMismatchMessage: (ConvertingMismatch) -> Unit = { message: ConvertingMismatch ->
+    System.err.println(message)
+}
+
+private val logger = LoggerFactory.getLogger(ArrowWriter::class.java)
+
+public val logMismatchMessage: (ConvertingMismatch) -> Unit = { message: ConvertingMismatch ->
+    logger.debug(message.toString())
+}
+
+/**
+ * Save [dataFrame] content in Apache Arrow format (can be written to File, ByteArray, OutputStream or raw Channel) with [targetSchema].
+ * If [dataFrame] content does not match with [targetSchema], behaviour is specified by [mode], mismatches would be sent to [mismatchSubscriber]
+ */
+public interface ArrowWriter : AutoCloseable {
+    public val dataFrame: DataFrame<*>
+    public val targetSchema: Schema
+    public val mode: Mode
+    public val mismatchSubscriber: (ConvertingMismatch) -> Unit
+
+    public companion object {
+
+        public fun create(
+            dataFrame: AnyFrame,
+            targetSchema: Schema,
+            mode: Mode,
+            mismatchSubscriber: (ConvertingMismatch) -> Unit = ignoreMismatchMessage
+        ): ArrowWriter = ArrowWriterImpl(dataFrame, targetSchema, mode, mismatchSubscriber)
+
+        /**
+         * If [restrictWidening] is true, [dataFrame] columns not described in [targetSchema] would not be saved (otherwise, would be saved as is).
+         * If [restrictNarrowing] is true, [targetSchema] fields that are not nullable and do not exist in [dataFrame] will produce exception (otherwise, would not be saved).
+         * If [strictType] is true, [dataFrame] columns described in [targetSchema] with non-compatible type will produce exception (otherwise, would be saved as is).
+         * If [strictNullable] is true, [targetSchema] fields that are not nullable and contain nulls in [dataFrame] will produce exception (otherwise, would be saved as is with nullable = true).
+         */
+        public class Mode(
```

Comment:
maybe make it a data class, better for equality and you gain .copy() functions for when someone wants STRICT but with one thing as false

Reply:
Done